pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,392 @@
+	}
+
+	s.kubeApiserverStaticPod = pod
+
+	return s.kubeApiserverStaticPod, nil
+}
+
+func (s *Apiserver) GetResourceDirPath() (string, error) {
+	if s.ResourceDirPath != "" {
+		return s.ResourceDirPath, nil
+	}
+
+	pod, err := s.GetCurrentApiserverManifest()
+	if err != nil {
+		return "", fmt.Errorf("failed to read apiserver pod manifest: %v", err)
+	}
+
+	resourceDir, err := GetVolumeHostPathPath("resource-dir", pod.Spec.Volumes)
+	if err != nil {
+		return "", fmt.Errorf("failed to find resource-dir: %v", err)
+	}
+
+	s.ResourceDirPath = *resourceDir
+
+	return s.ResourceDirPath, nil
+}
+
+func (s *Apiserver) GetRecoveryDir() string {
+	return s.recoveryDir
+}
+
+func (s *Apiserver) RestConfig() (*rest.Config, error) {
+	if s.restConfig == nil {
+		return nil, errors.New("no rest config is set yet")
+	}
+
+	return s.restConfig, nil
+}
+
+func (s *Apiserver) KubeConfig() (*clientcmdapiv1.Config, error) {
+	restConfig, err := s.RestConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return &clientcmdapiv1.Config{
+		APIVersion: "v1",
+		Clusters: []clientcmdapiv1.NamedCluster{
+			{
+				Name: "recovery",
+				Cluster: clientcmdapiv1.Cluster{
+					CertificateAuthority: restConfig.CAFile,
+					Server:               restConfig.Host,
+				},
+			},
+		},
+		Contexts: []clientcmdapiv1.NamedContext{
+			{
+				Name: "admin",
+				Context: clientcmdapiv1.Context{
+					Cluster:  "recovery",
+					AuthInfo: "admin",
+				},
+			},
+		},
+		CurrentContext: "admin",
+		AuthInfos: []clientcmdapiv1.NamedAuthInfo{
+			{
+				Name: "admin",
+				AuthInfo: clientcmdapiv1.AuthInfo{
+					ClientCertificateData: restConfig.CertData,
+					ClientKeyData:         restConfig.KeyData,
+				},
+			},
+		},
+	}, nil
+}
+
+func (s *Apiserver) GetKubeClientset() (*kubernetes.Clientset, error) {
+	if s.kubeClientSet != nil {
+		return s.kubeClientSet, nil
+	}
+
+	restConfig, err := s.RestConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	kubeClientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build kubernetes clientset: %v", err)
+	}
+
+	s.kubeClientSet = kubeClientset
+
+	return s.kubeClientSet, nil
+}
+
+func (s *Apiserver) recoveryPod() (*corev1.Pod, error) {
+	// Create the manifest to run recovery apiserver
+	recoveryPodTemplateBytes, err := v311_00_assets.Asset(RecoveryPodAsset)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find internal recovery pod asset %q: %v", RecoveryPodAsset, err)
+	}
+
+	// Process the template
+	t, err := template.New("recovery-pod-template").Parse(string(recoveryPodTemplateBytes))
+	if err != nil {
+		return nil, fmt.Errorf("fail to parse internal recovery pod template %q: %v", RecoveryPodAsset, err)
+	}
+
+	recoveryPodBuffer := bytes.NewBuffer(nil)
+	err = t.Execute(recoveryPodBuffer, struct {
+		KubeApiserverImage string
+		ResourceDir        string
+	}{
+		KubeApiserverImage: s.KubeApiserverImage,
```

Comment:
I expected you to get this value from the existing kube-apiserver pod manifest that you're already reading.

Reply:
done